adam_preprocessing/shape_stroke_extraction.py

```diff
@@ -111,6 +111,20 @@ def kp2stroke(strokeinfo):
     return connection
 
 
+def objects_touching(object_one, object_two):
+    """Determine if two objects should be considered 'touching' based on the
+    euclidean (absolute) distance of their closest two points.
+    """
+    object_one_points = np.concatenate(object_one['stroke_graph']['original_reduced_strokes'], axis=0)
+    object_two_points = np.concatenate(object_two['stroke_graph']['original_reduced_strokes'], axis=0)
+    for point_one in object_one_points:
+        for point_two in object_two_points:
+            distance = np.sum((point_one - point_two) ** 2) ** 0.5
+            if distance < TOUCHING_DISTANCE_THRESHOLD:
+                return True
+    return False
```

Comment:
IIRC NumPy performs better when using its whole-array operations instead of Python loops. If so we might speed this up by switching to a loop-free version. This doesn't handle shapes correctly, but something like this:

```python
return np.any(
  np.sum(
    (object_one_points[:, None] - object_two_points) ** 2
  ) ** 0.5 < TOUCHING_DISTANCE_THRESHOLD
)
```

Reply:
Testing with some arrays in the REPL, this sum should work (I think) but we'll need `axis=-1`. I'm assuming the concatenated arrays have shape `(10 * n_strokes, 3)`. Toy example:

```python
import numpy as np
x = np.array(
    [
        [0, 0, 1],
        [1, 0, 0],
    ]
)
y = np.array(
    [
        [0, 0, 2],
        [5, 0, 0],
        [0, 9, 0],
    ]
)
distances = np.sum((x[:, None] - y) ** 2, axis=-1) ** 0.5
```

Reply:
This looks good to me (with the `axis=-1` for the `sum()` to just combine coordinates of course)!

I had thought about this, too, but I couldn't think of how to do it but I figured the cost in human time of looking up how to do this in NumPy wasn't worth the saved computer time. Since you wrote it up, I'll incorporate it.

Reply:
That's fair. I too worried that this might be too gross to be worthwhile, but it turned out easier than expected when I looked into it.